[case testNarrowToLiteral]
a = "a"
reveal_type(a)  # N: Revealed type is "'a'" (narrowed from "str")
a = "b"
reveal_type(a)  # N: Revealed type is "'b'" (narrowed from "str")
b = a
reveal_type(b)  # N: Revealed type is "'b'" (narrowed from "str")
b = "c"
reveal_type(b)  # N: Revealed type is "'c'" (narrowed from "str")


[case testNarrowToLiteralGeneric]
from helper import T
def f(t: T) -> T: ...

a = f(1)
reveal_type(a)  # N: Revealed type is "1" (narrowed from "int")
